# Spatial domain identification: computational methods for discovering tissue architecture

## Introduction
This workshop will focus on computational approaches for identifying and characterizing spatial domains in single-cell spatial transcriptomics (ST) data. Spatial domains are tissue regions that share similar features, such as similar gene expression profiles and cell type abundances. For analysis of ST data, it is generally necessary to identify these domains to understand their dynamics under different tissue conditions, such as between healthy and disease states.

As ST data is increasingly used in research due to the benefit of in-situ identification of transcripts and cells, several computational approaches for spatial domain identification have been developed. In this workshop, participants will learn methods for unsupervised detection of spatial domains with distinct molecular signatures and understand techniques for biological interpretation of spatial domains along with associated caveats. Through hands-on tutorials, participants will learn about and apply state-of-the-art domain identification algorithms to real spatial transcriptomics datasets.

## Provisional schedule:

### Session 1 - Introduction to spatial domain identification
### Session 2 - Methodological foundations
### Session 3 - Preprocessing, implementation and evaluation metrics
### Session 4 - Analysis and biological interpretation


## Data
We'll use a subset of Xenium kidney data from [Spatio-temporal interaction of immune and renal cells determines glomerular crescent formation in autoimmune kidney disease](https://www.biorxiv.org/content/10.1101/2024.12.18.629206v1)